### PR TITLE
[10156] Capture sys.stdout from disttrial worker sub-process.

### LINKED ISSUE #10156: distrial  - Capture sys.stdout from sub-process
|[adiroiban's avatar](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#10156|
|Type|enhancement|
|Created|2021-03-27 21:17:26Z|

The distributed sub-process worker is collecting and forwarding to the central process only the twisted.log logs.

It would be nice to also capture and forwadd all sys.stdout.

The idea is that since `--debug` is not supported in distrial, sys.stdout provides a quick way to debug  the sub-process.

Even if you use remote pdb, it would still help to get the sys.stout to see for example the port number on which the remote pdb is listening.

This is only for capturing the high-level sys.stdout access and not the low-level fd fileno=1.

<details><summary>Searchable metadata</summary>

```
trac-id__10156 10156
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__trial trial
keywords__None None
time__1616879846056100 1616879846056100
changetime__1626936120366623 1626936120366623
version__None None
owner__adiroiban adiroiban

```
</details>

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Fixes #10156

This updates the log capturing of the distributed sub-process to also capture the data written to `sys.stdout`.

This is a proof of concept for a future PR in which the fd=1 (sys.stdout) will be used for AMP protocol communication to make it easier to run distrial on windows.

For now it reuses the existing `testWrite` which is documented as `Print test output from the worker.`

But in trunk `testWrite` was more like `Print python.log output from the worker.`.

This only handles the high-level sys.stdout writing.

I find it easier to check the output of a test in a single file, rather than checking both `N/test.log` and `N/out.log`.

In an ideal case, I would like to have a single test.log file that will centralize all the logs from the sub-processes.

A sub-process that writes directly to FD=1 will continue to have the data written to `_trial_temp/N/out.log`
A sub-process that writes to stderr will continue to have the data written to `_trial_temp/N/err.log` just like in trunk.

## Discussion

* See if `testWrite` should be renamed to `writeTestOut` or `writeTestLog` or  `writeOut` and have it dedicated to capturing the output generated via python log system
* Add a new `writeTestStdout` to have it dedicated for capturing sys.stdout and fd=1 data from sub-process
* Add a new `writeTestStderr` to have it dedicated for capturing sys.stderr and fd=2 data from sub-process

My preference:
* Merging log and stdout data in a single stream via a renamed `testWrite` to `writeOut`, but have a separate stderr stream.
* Have a single test.log file that aggregates the logs (python.log and sys.stdout) from all sub-processes 
* We can still have  `_trial_temp/N/out.log` and `_trial_temp/N/err.log` but they are handled in the sub-process. Not in the main control process like we have now. 
* Fail if a test tries to write the stdout. I would argue that tests  writing to stdout are a bad practice as the output is most probably not observed under CI.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10156
* [NO] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: adiroiban
Reviewer: 
Fixes: ticket:10156

Capture sys.stdout from distrial worker sub-processes and store it in the centralized log file.
```

### CODE REVIEW

**Comment by adiroiban** [comment]
I have no executed `tox -e lint` as it fails and I don't have the energy now to debug tox... I will let the CI do its job and help me with that.

I did manually run `pyflakes` and `mypy`

```
$ tox --version
3.21.4 imported from /usr/lib/python3/dist-packages/tox/__init__.py
✔  ~/chevah/twisted [10156-distrial-sys-stdout|⚑ 2] 

22:56 $ tox -e lint
.tox create: /home/adi/chevah/twisted/.tox/.tox
ERROR: invocation failed (exit code 1), logfile: /home/adi/chevah/twisted/.tox/.tox/log/.tox-0.log
============================================================ log start ============================================================
RuntimeError: failed to build image setuptools because:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/virtualenv/seed/embed/via_app_data/via_app_data.py", line 57, in _install
    installer.install(creator.interpreter.version_info)
  File "/usr/lib/python3/dist-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 42, in install
    self._sync(filename, into)
  File "/usr/lib/python3/dist-packages/virtualenv/seed/embed/via_app_data/pip_install/copy.py", line 13, in _sync
    copy(src, dst)
  File "/usr/lib/python3/dist-packages/virtualenv/util/path/_sync.py", line 53, in copy
    method(norm(src), norm(dest))
  File "/usr/lib/python3/dist-packages/virtualenv/util/path/_sync.py", line 64, in copytree
    shutil.copy(src_f, dest_f)
  File "/usr/lib/python3.9/shutil.py", line 419, in copy
    copymode(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.9/shutil.py", line 308, in copymode
    chmod_func(dst, stat.S_IMODE(st.st_mode))
FileNotFoundError: [Errno 2] No such file or directory: '/home/adi/chevah/twisted/.tox/.tox/lib/python3.9/site-packages/pkg_resources/_vendor/packaging/_typing.py'


============================================================= log end =============================================================
ERROR: InvocationError for command /usr/bin/python3 -m virtualenv --no-download --python /usr/bin/python3 .tox (exited with code 1)
/home/adi/chevah/twisted/.tox/.tox/bin/python: No module named tox
```

File: src/twisted/trial/_dist/test/test_workertrial.py [hunk 1]
```diff
@@ -40,16 +57,57 @@ def test_emit(self):
...
+        the centralized log.
+
+        It's kind of a manual test, as there is no automatic assertion.
+        When call as
+        C{trial -j1 twisted.trial._dist.test.test_workertrial.WorkerStdoutTests.test_integration}
+
+        You should see the message via C{cat _trial_temp/0/test.log}
+        """
+        sys.stdout.write("Hello from \N{sun}y worker!")
+        sys.stdout.flush()
```
Comment (adiroiban) [inline src/twisted/trial/_dist/test/test_workertrial.py:108]: I have no idea how to to test this in other way.

There is a unit test for `main()` but I also want to have a test in which the real code is exercised.

**Comment by adiroiban** [comment]
This is ready for an initial review.

At this stage the code is POC to support the future work for getting distributed trial on Windows.


I don't know why mypy is not happy with

```
src/twisted/trial/_dist/workertrial.py:143:9: error: Statement is unreachable  [unreachable]
src/twisted/trial/_dist/test/test_workertrial.py:32:1: error: Cannot determine consistent method resolution order (MRO) for "FakeAM
```

I have check the line and with `$ trial --coverage -j1 twisted.trial._dist.test.test_options` the line is reached.
Also, I have no idea why is not happy with `class FakeAMP(AMP):`

**Review by glyph — CHANGES_REQUESTED** [review]
I think that this is a bad strategy.

One of the things I've learned from working with Python process pools is that one big reason to have a process pool is to isolate a sometimes-misbehaving C extension, and one of the ways that misbehaving C extensions can misbehave is by emitting messages *directly* to sys.stderr or sys.stdout.

I think it's important that, when we can, on UNIX at least (and maybe on Windows if we can sort out the file descriptor / standard I/O shenanigans?  maybe use a named pipe or something, Windows XP is dead, right?) we should stick to defaulting to capturing sys.stdout by using a non-stdio FD for inter-process communication.  If Windows really can't do that, then let's leave this code around.

**Comment by graingert** [comment]
> Windows XP is dead, right?

Yep cpython 3.4 and 2.7 were last releases to support it

**Comment by adiroiban** [comment]
Thanks for the review. I don't plan to work on this in the near future.

I have limited time so my plan was to get distributed trial on windows using the option that will require less effort.

**Pull request closed without merging**